in_toto/runlib.go

```diff
@@ -299,14 +309,71 @@ func InTotoRun(name string, runDir string, materialPaths []string, productPaths
 		Command:     cmdArgs,
 		Environment: map[string]interface{}{},
 	}
+
 	linkMb.Signatures = []Signature{}
 	// We use a new feature from Go1.13 here, to check the key struct.
 	// IsZero() will return True, if the key hasn't been initialized
+
 	// with other values than the default ones.
 	if !reflect.ValueOf(key).IsZero() {
 		if err := linkMb.Sign(key); err != nil {
 			return linkMb, err
 		}
 	}
+
+	return linkMb, nil
+}
+
+func InTotoRecordStart(name string, materialPaths []string, key Key, hashAlgorithms, gitignorePatterns []string, lStripPaths []string) (Metablock, error) {
+	var linkMb Metablock
+	materials, err := RecordArtifacts(materialPaths, hashAlgorithms, gitignorePatterns, lStripPaths)
+	if err != nil {
+		return linkMb, err
+	}
+
+	linkMb.Signed = Link{
+		Type:        "link",
+		Name:        name,
+		Materials:   materials,
+		Products:    map[string]interface{}{},
+		ByProducts:  map[string]interface{}{},
+		Command:     []string{},
+		Environment: map[string]interface{}{},
+	}
+
+	if !reflect.ValueOf(key).IsZero() {
```

Comment:
Why not just a key == nil check on that position? I see this reflect isZero construct the first time. It checks for zero values in the key object right? Like: The key has been initialized but it's empty? Is this correct?

Reply:
Correct, this is checking to see if the value of `key` is the zero value of the `Key` struct.  A nil check won't work here due to `key` being an instance of `Key` and not `*Key`.

An equivalent way to write this would be `!key == Key{}`, which may be a bit more clear.  This was done this way to remain consistent with the check elsewhere in the code base (specifically here https://github.com/in-toto/in-toto-golang/pull/119/files/31b889b97d898f343cb7c6b3c1ef092dbe5f5506#diff-20ae90a40c9494c3e2e4a995a57bc508621a0e473dd74ae811f560d187ce7887R318 )

Reply:
Thank you for the explanation :)  I learned something new.